### DISCUSSION TITLE
Check if commit is inside branch

### DISCUSSION THREAD

**Discussion by fchastanet** [community/community#24660] [Apps, API and Webhooks]
<p>Hello</p>
<p>I’m trying to find a better way to check if a commit is inside a branch</p>
<p>the only solution I found is to use compare method from api v3</p>
<p><a href="https://developer.github.com/v3/repos/commits/#compare-two-commits" rel="nofollow noopener">https://developer.github.com/v3/repos/commits/#compare-two-commits</a></p>
<p>the only problem is that it will return all the diffs, which I don’t want, it could lead to performance issue for large repos.</p>
<p>I tried to find a way with graphql too but without success.</p>
<p>Have you an idea to how return just the ahead/behind stats </p>
<p>or just true/false status like done with</p>
<pre><code>git branch --contains $COMMIT_ID
</code></pre>
<p>?</p>
<p>kind regards</p>

**Comment by fchastanet** [comment]
<p>does anyone has a solution for this ?</p>

**Reply by ivasilyeu** [reply]
How about this:
if [ -z "$(git cherry -v main $COMMIT_HASH)" ]; then
    echo "Commit exists on the main branch."
else
    echo "Commit does not exist on the main branch."
fi

**Comment by kabbottcc** [comment]
<p>FWIW, I came here looking for the same answer. Is the only available option via CLI?</p>

**Comment by fchastanet** [comment, marked as answer]
<p>finally I used API v3 REST for this use case<br>
graphql is not good for recursive<br>
both APIs are complementay</p>

**Reply by bradyclifford** [reply]
You mind sharing your solution?

**Reply by fchastanet** [reply]
Hello of course, it is php code

The real question for me was to know if the master was merged in the branch
I compute it like this in the code below
```php
$graphql['featureBranchContainsMasterLastCommit'] = ($compare['behind_by'] === 0 && $compare['status'] !== 'diverged');
```

You just need a GITHUB_TOKEN constant which must remains secret of course
You have to set the variables $owner, $repo, $defaultRepoMasterBranch, $featureBranchOid
The `$compare` object contains the response of github in json format
But check [this documentation](https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28#compare-two-commits) as the github API  probably changed since the time I wrote this code
```php
/**
 * @param string $method
 * @param string $uri
 * @param array $headers
 * @param array $data
 * @return mixed
 * @throws \GuzzleHttp\Exception\GuzzleException
 */
function guzzleGithubApi(string $method, string $uri, array $headers = [], array $data = []) {
    if (!isset($headers['Authorization'])) {
        $headers['Authorization'] = 'token ' . GITHUB_TOKEN;
    }
    $client = new GuzzleHttp\Client(['headers' => $headers]);
    $response = $client->request(
        $method,
        $uri,
        $data
    );

    $responseJson = json_decode($response->getBody(), true);
    if (isset($responseJson['errors'])) {
        error_log("api error $uri ".$response->getStatusCode() . " message " . $response->getBody());
    }
    return $responseJson;
}
try {
    $compare = guzzleGithubApi(
        'GET',
        "https://api.github.com/repos/$owner/$repo/compare/$defaultRepoMasterBranch...$featureBranchOid",
        ['Accept' => 'application/vnd.github.v3+json, application/vnd.github.sailor-v-preview+json']
    );
    // enrich graphql response
    $graphql['featureBranchCompareMaster'] = $compare;
    $graphql['featureBranchContainsMasterLastCommit'] = ($compare['behind_by'] === 0 && $compare['status'] !== 'diverged');
} catch (\GuzzleHttp\Exception\GuzzleException $ex) {
    error_log("{$ex->getMessage()} - {$ex->getTraceAsString()}");
    // ignore
}
```

**Comment by michaljonny** [comment]
<p>In Git, a branch is a pointer to one specific commit, while a commit is a snapshot of your repository at a specific point in time. Your branch pointer moves along with each new commit you make.</p>

**Comment by dennyjon** [comment]
let me know if you get the ans[.](payslips-plus.co.uk) thanks
